From 00ff0c97de90cd83a89e475069a590adb1c32c1e Mon Sep 17 00:00:00 2001
From: Roger Siddons <dizygotheca@ntlworld.com>
Date: Sun, 25 Jan 2015 00:16:01 +0000
Subject: [PATCH 1/4] Centre zoomed images

Images that are smaller than a widget are centred, but the image zoom is currently ignored. This results in a slideshow image being displaced when it is zoomed.
This patch improves the problem but it may not be a complete solution as the image still wobbles when being zoomed.

diff --git a/mythtv/libs/libmythui/mythuianimation.cpp b/mythtv/libs/libmythui/mythuianimation.cpp
index 2b66467..39fbec7 100644
--- a/mythtv/libs/libmythui/mythuianimation.cpp
+++ b/mythtv/libs/libmythui/mythuianimation.cpp
@@ -4,6 +4,47 @@
 
 #include <QDomDocument>
 
+QRect UIEffects::GetExtent(const QSize &size)
+{
+    int x = 0, y = 0;
+    int zoomedWidth = size.width() * hzoom;
+    int zoomedHeight = size.height() * vzoom;
+
+    switch (centre)
+    {
+    case TopLeft:
+    case Top:
+    case TopRight:
+        y = -zoomedHeight / 2; break;
+    case Left:
+    case Middle:
+    case Right:
+        y = -size.height() / 2; break;
+    case BottomLeft:
+    case Bottom:
+    case BottomRight:
+        y = size.height() - zoomedHeight / 2; break;
+    }
+
+    switch (centre)
+    {
+    case TopLeft:
+    case Left:
+    case BottomLeft:
+        x = -zoomedWidth / 2; break;
+    case Top:
+    case Middle:
+    case Bottom:
+        x = -size.width() / 2; break;
+    case TopRight:
+    case Right:
+    case BottomRight:
+        x = size.width() - zoomedWidth / 2; break;
+    }
+
+    return QRect(x, y, zoomedWidth, zoomedHeight);
+}
+
 MythUIAnimation::MythUIAnimation(MythUIType* parent, Trigger trigger, Type type)
     : m_parent(parent), m_type(type), m_trigger(trigger),
       m_centre(UIEffects::Middle), m_active(false), m_looped(false),
diff --git a/mythtv/libs/libmythui/mythuianimation.h b/mythtv/libs/libmythui/mythuianimation.h
index 3c605da..9afa361 100644
--- a/mythtv/libs/libmythui/mythuianimation.h
+++ b/mythtv/libs/libmythui/mythuianimation.h
@@ -31,6 +31,8 @@ class UIEffects
         return QPointF(x, y);
     }
 
+    QRect GetExtent(const QSize &size);
+
     int    alpha;
     float  hzoom;
     float  vzoom;
diff --git a/mythtv/libs/libmythui/mythuiimage.cpp b/mythtv/libs/libmythui/mythuiimage.cpp
index 1637b2a..03dcc80 100644
--- a/mythtv/libs/libmythui/mythuiimage.cpp
+++ b/mythtv/libs/libmythui/mythuiimage.cpp
@@ -1278,17 +1278,17 @@ void MythUIImage::DrawSelf(MythPainter *p, int xoffset, int yoffset,
         if (!m_imageProperties.forceSize.isNull())
             area.setSize(area.size().expandedTo(currentImage->size()));
 
-        // Centre image in available space
-        int x = 0;
-        int y = 0;
+        // Centre image in available space, accounting for zoom
+        int x = 0, y = 0;
+        QRect visibleImage = m_Effects.GetExtent(currentImageArea.size());
 
-        if (area.width() > currentImageArea.width())
-            x = (area.width() - currentImageArea.width()) / 2;
+        if (area.width() > visibleImage.width())
+            x = area.width() / 2 + visibleImage.topLeft().x();
 
-        if (area.height() > currentImageArea.height())
-            y = (area.height() - currentImageArea.height()) / 2;
+        if (area.height() > visibleImage.height())
+            y = area.height() / 2 + visibleImage.topLeft().y();
 
-        if (x > 0 || y > 0)
+        if ((x > 0 || y > 0))
             area.translate(x, y);
 
         QRect srcRect;
-- 
1.9.1

